Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs

```diff
@@ -458,34 +540,84 @@ public class DafnyModel {
           utfCode = ((Model.Integer)fCharToInt.OptEval(elt)).AsInt();
           return "'" + char.ConvertFromUtf32(utfCode) + "'";
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnreservedCharValue(elt);
       }
       if (fType.OptEval(elt) == fReal.GetConstant()) {
         if (fU2Real.OptEval(elt) != null) {
           return CanonicalName(fU2Real.OptEval(elt));
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnreservedNumericValue(elt, Type.Real);
       }
       if (fType.OptEval(elt) == fBool.GetConstant()) {
         if (fU2Bool.OptEval(elt) != null) {
           return CanonicalName(fU2Bool.OptEval(elt));
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnreservedBoolValue(elt);
       }
       if (fType.OptEval(elt) == fInt.GetConstant()) {
         if (fU2Int.OptEval(elt) != null) {
           return CanonicalName(fU2Int.OptEval(elt));
         }
-        return "?#" + GetShortElementId(elt);
+        return GetUnreservedNumericValue(elt, Type.Int);
       }
       return "";
     }
 
-    private int GetShortElementId(Model.Element element) {
-      if (!shortElementIds.ContainsKey(element.Id)) {
-        shortElementIds[element.Id] = shortElementIds.Count;
+    /// <summary>
+    /// Find a char value that is different from any other value
+    /// of that type in the entire model. Reserve that value for given element
+    /// </summary>
+    private string GetUnreservedCharValue(Model.Element element) {
+      if (reservedValuesMap.ContainsKey(element)) {
+        return reservedValuesMap[element];
+      }
+      int i = firstCharacterUtfValue;
+      while (reservedChars.Contains(Convert.ToChar(i))) {
+        i++;
+      }
+      reservedValuesMap[element] = $"'{Convert.ToChar(i)}'";
+      reservedChars.Add(Convert.ToChar(i));
+      return reservedValuesMap[element];
+    }
+
+    /// <summary>
+    /// Find a bool value that is different from any other value
+    /// of that type in the entire model (if possible).
+    /// Reserve that value for given element
+    /// </summary>
+    private string GetUnreservedBoolValue(Model.Element element) {
+      if (reservedValuesMap.ContainsKey(element)) {
```

Comment:
`reservedValuesMap.TryGetValue(element, out var reservedValue)` might be handy here. Should work for all other `GetUnreserved*` methods too.

Reply:
This does look much nicer! Thank you, changed this in the last commit